### ISSUE TITLE
IndexOutOfBoundsException in LinearLayout using tk.zielony:carbon:0.14.0

### ISSUE THREAD

**Issue by NeverGivinUp** [ZieIony/Carbon#316]
I'm getting much more crash reports, since Google changed it's reporting policy. This is from a app version I published two days ago. Unfortunately I don't know which activity and fragment the layout is in. If there is a bug in my code, the layout should report it more early and with some context. Otherwise, please look into this bug. 

Thanks a lot!

Sony Xperia XZ (F8331), 3072MB RAM, Android 7.1
```
java.lang.IndexOutOfBoundsException: 
at java.util.ArrayList.get (ArrayList.java:411)
at carbon.widget.LinearLayout.getChildDrawingOrder (LinearLayout.java:329)
at android.view.View.populateAccessibilityNodeInfoDrawingOrderInParent (View.java:7082)
at android.view.View.onInitializeAccessibilityNodeInfoInternal (View.java:7041)
at android.view.ViewGroup.onInitializeAccessibilityNodeInfoInternal (ViewGroup.java:3069)
at android.view.View$AccessibilityDelegate.onInitializeAccessibilityNodeInfo (View.java:23621)
at android.support.v4.view.AccessibilityDelegateCompatIcs.onInitializeAccessibilityNodeInfo (AccessibilityDelegateCompatIcs.java:101)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateIcsImpl.onInitializeAccessibilityNodeInfo (AccessibilityDelegateCompat.java:198)
at android.support.v4.view.AccessibilityDelegateCompat.onInitializeAccessibilityNodeInfo (AccessibilityDelegateCompat.java:451)
at android.support.v7.widget.RecyclerViewAccessibilityDelegate.onInitializeAccessibilityNodeInfo (RecyclerViewAccessibilityDelegate.java:56)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateJellyBeanImpl$1.onInitializeAccessibilityNodeInfo (AccessibilityDelegateCompat.java:246)
at android.support.v4.view.AccessibilityDelegateCompatJellyBean$1.onInitializeAccessibilityNodeInfo (AccessibilityDelegateCompatJellyBean.java:66)
at android.view.View.onInitializeAccessibilityNodeInfo (View.java:6655)
at android.view.View.createAccessibilityNodeInfoInternal (View.java:6616)
at android.view.View$AccessibilityDelegate.createAccessibilityNodeInfo (View.java:23690)
at android.view.View.createAccessibilityNodeInfo (View.java:6599)
at android.view.accessibility.AccessibilityRecord.setSource (AccessibilityRecord.java:145)
at android.view.accessibility.AccessibilityRecord.setSource (AccessibilityRecord.java:119)
at android.view.View.onInitializeAccessibilityEventInternal (View.java:6553)
at android.view.View$AccessibilityDelegate.onInitializeAccessibilityEvent (View.java:23602)
at android.support.v4.view.AccessibilityDelegateCompatIcs.onInitializeAccessibilityEvent (AccessibilityDelegateCompatIcs.java:97)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateIcsImpl.onInitializeAccessibilityEvent (AccessibilityDelegateCompat.java:192)
at android.support.v4.view.AccessibilityDelegateCompat.onInitializeAccessibilityEvent (AccessibilityDelegateCompat.java:432)
at android.support.v7.widget.RecyclerViewAccessibilityDelegate.onInitializeAccessibilityEvent (RecyclerViewAccessibilityDelegate.java:65)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateJellyBeanImpl$1.onInitializeAccessibilityEvent (AccessibilityDelegateCompat.java:241)
at android.support.v4.view.AccessibilityDelegateCompatJellyBean$1.onInitializeAccessibilityEvent (AccessibilityDelegateCompatJellyBean.java:61)
at android.view.View.onInitializeAccessibilityEvent (View.java:6539)
at android.view.View.sendAccessibilityEventUncheckedInternal (View.java:6406)
at android.view.View$AccessibilityDelegate.sendAccessibilityEventUnchecked (View.java:23541)
at android.support.v4.view.AccessibilityDelegateCompatIcs.sendAccessibilityEventUnchecked (AccessibilityDelegateCompatIcs.java:122)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateIcsImpl.sendAccessibilityEventUnchecked (AccessibilityDelegateCompat.java:223)
at android.support.v4.view.AccessibilityDelegateCompat.sendAccessibilityEventUnchecked (AccessibilityDelegateCompat.java:371)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateJellyBeanImpl$1.sendAccessibilityEventUnchecked (AccessibilityDelegateCompat.java:268)
at android.support.v4.view.AccessibilityDelegateCompatJellyBean$1.sendAccessibilityEventUnchecked (AccessibilityDelegateCompatJellyBean.java:87)
at android.view.View.sendAccessibilityEventUnchecked (View.java:6389)
at android.support.v7.widget.RecyclerView.sendAccessibilityEventUnchecked (RecyclerView.java:3206)
at android.view.View.sendAccessibilityEventInternal (View.java:6368)
at android.view.View$AccessibilityDelegate.sendAccessibilityEvent (View.java:23499)
at android.support.v4.view.AccessibilityDelegateCompatIcs.sendAccessibilityEvent (AccessibilityDelegateCompatIcs.java:117)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateIcsImpl.sendAccessibilityEvent (AccessibilityDelegateCompat.java:217)
at android.support.v4.view.AccessibilityDelegateCompat.sendAccessibilityEvent (AccessibilityDelegateCompat.java:349)
at android.support.v4.view.AccessibilityDelegateCompat$AccessibilityDelegateJellyBeanImpl$1.sendAccessibilityEvent (AccessibilityDelegateCompat.java:263)
at android.support.v4.view.AccessibilityDelegateCompatJellyBean$1.sendAccessibilityEvent (AccessibilityDelegateCompatJellyBean.java:82)
at android.view.View.sendAccessibilityEvent (View.java:6333)
at android.view.View.onFocusChanged (View.java:6272)
at android.view.View.handleFocusGainInternal (View.java:6013)
at android.view.ViewGroup.handleFocusGainInternal (ViewGroup.java:717)
at android.view.View.requestFocusNoSearch (View.java:9178)
at android.view.View.requestFocus (View.java:9157)
at android.view.ViewGroup.requestFocus (ViewGroup.java:2835)
at android.view.ViewGroup.onRequestFocusInDescendants (ViewGroup.java:2875)
at android.view.ViewGroup.requestFocus (ViewGroup.java:2831)
at android.view.ViewGroup.onRequestFocusInDescendants (ViewGroup.java:2875)
at android.view.ViewGroup.requestFocus (ViewGroup.java:2831)
at android.view.ViewGroup.onRequestFocusInDescendants (ViewGroup.java:2875)
at android.view.ViewGroup.requestFocus (ViewGroup.java:2831)
at android.view.ViewGroup.onRequestFocusInDescendants (ViewGroup.java:2875)
at android.view.ViewGroup.requestFocus (ViewGroup.java:2831)
at android.view.ViewGroup.onRequestFocusInDescendants (ViewGroup.java:2875)
at android.view.ViewGroup.requestFocus (ViewGroup.java:2831)
at android.view.ViewGroup.onRequestFocusInDescendants (ViewGroup.java:2875)
at android.view.ViewGroup.requestFocus (ViewGroup.java:2834)
at android.view.View.requestFocus (View.java:9124)
at android.view.ViewRootImpl.performTraversals (ViewRootImpl.java:2150)
at android.view.ViewRootImpl.doTraversal (ViewRootImpl.java:1254)
at android.view.ViewRootImpl$TraversalRunnable.run (ViewRootImpl.java:6343)
at android.view.Choreographer$CallbackRecord.run (Choreographer.java:874)
at android.view.Choreographer.doCallbacks (Choreographer.java:686)
at android.view.Choreographer.doFrame (Choreographer.java:621)
at android.view.Choreographer$FrameDisplayEventReceiver.run (Choreographer.java:860)
at android.os.Handler.handleCallback (Handler.java:751)
at android.os.Handler.dispatchMessage (Handler.java:95)
at android.os.Looper.loop (Looper.java:241)
at android.app.ActivityThread.main (ActivityThread.java:6274)
at java.lang.reflect.Method.invoke (Method.java)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run (ZygoteInit.java:886)
at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:776)
```

**Comment by ZieIony** [comment]
Index of a child is being retrieved before the view list is build. A fix should be as simple as overriding this method in all layouts:

```Java
@Override
protected int getChildDrawingOrder(int childCount, int child) {
    if (views.size() != childCount)
        getViews();    // rebuild the list
    return indexOfChild(views.get(child));
}
```

**Comment by NeverGivinUp** [comment]
If that's all that's necessary, that would be great. 
If version 0.15 is not ready for release yet (e.g. it might be insufficiently tested) would you kindly prepare a version 0.14.1 with just the fix.

**Comment by ZieIony** [comment]
0.15.0 is fine, but I'm afraid that you would have to update a lot more than just the version number. If you wish, you can try the most recent commit using JitPack. It contains the fix and will be released as 0.15.0. That should give you some insight on what has changed between 0.14.0 and 0.15.0.

**Comment by NeverGivinUp** [comment]
I don't mind API changes. But are there any changes that I need to adapt to, which the compiler won't catch?

**Comment by ZieIony** [comment]
0.15.0 is on Maven. The non-API changes are mostly in styles. Some colors may change.

**Comment by NeverGivinUp** [comment]
I've upgraded to 0.15.0 and also upgraded Android Studio to 3.0 beta 2, because 0.15.0 requires Java 8. After adapting some code it builds OK.

Unfortunately I can't get it to generate an APK. Whatever I do -- using or disallowing MultiDex; excluding transiently included module dependencies, like most android support modules; enabling or disabling Aapt2; cleaning, deleting the build folder, invalidating the android caches and restarting it as well and rebuilding -- I always get an exception like this one:

```
:gastricBypassSupport:transformDexArchiveWithDexMergerForDebug
AGPBI: {"kind":"error","text":"Error converting bytecode to dex:\nCause: com.android.dex.DexException: Multiple dex files define Landroid/support/v8/renderscript/Matrix2f;","sources":[{}],"original":"UNEXPECTED TOP-LEVEL EXCEPTION:\ncom.android.dex.DexException: Multiple dex files define Landroid/support/v8/renderscript/Matrix2f;\n","tool":"Dex"}
AGPBI: {"kind":"error","text":"com.android.dex.DexException: Multiple dex files define Landroid/support/v8/renderscript/Matrix2f;","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat com.android.dx.merge.DexMerger.readSortableTypes(DexMerger.java:661)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat com.android.dx.merge.DexMerger.getSortedTypes(DexMerger.java:616)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat com.android.dx.merge.DexMerger.mergeClassDefs(DexMerger.java:598)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat com.android.dx.merge.DexMerger.mergeDexes(DexMerger.java:171)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat com.android.dx.merge.DexMerger.merge(DexMerger.java:198)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat com.android.builder.dexing.DexArchiveMergerCallable.call(DexArchiveMergerCallable.java:61)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat com.android.builder.dexing.DexArchiveMergerCallable.call(DexArchiveMergerCallable.java:36)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat java.util.concurrent.ForkJoinTask$AdaptedCallable.exec(ForkJoinTask.java:1445)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:289)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat java.util.concurrent.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:902)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1689)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1644)","sources":[{}]}
AGPBI: {"kind":"error","text":"\tat java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:157)","sources":[{}]}

 FAILED
```

The class that is defined by multiple dex files is often a different one, but always part of the renderscript support library. I don't know how to exclude the transient gradle dependency of the renderscript library, as it is not a maven module but something local, configured by Android.

Could this be a bug in the new gradle plug-in? Am I missing a gradle configuration? Any suggestions?

Thanks a lot!

**Comment by ZieIony** [comment]
Remove the RenderScript setup from your `app.gradle` file:

```
renderscriptTargetApi 20
renderscriptSupportModeEnabled true
```

Apparently the new gradle plugin includes RenderScript twice - from the library and from the app. I can't remove it from the library as it would fail to compile. That also means that hardware accelerated shadows are on by default. RenderScript never was easy to work with.

You can also exclude RenderScript on library's side:

https://stackoverflow.com/questions/41469772/exclude-package-classes-from-aar-gradle-dependency

**Comment by NeverGivinUp** [comment]
Thanks, that works now :-) I also saw, that you changed the installation instructions, but that alone wouldn't have saved me, because I wouldn't have looked, assuming that I already *have* installed carbon properly. A reminder to look at the release notes, update- or installation-instructions at the top of README.md, would draw my attention better.

One thing I noticed, is that interoperability of carbon and plain old android widgets changed. I used to have carbon radio buttons in a vanilla android radio group. That used to work before, but does not any more.
